File: Plugin/TimeLogPligin/src/main/java/com/johnson/asm/timelog/TimeLogPlugin.java
package com.johnson.asm.timelog;

import com.android.build.gradle.AppExtension;
import com.android.build.gradle.AppPlugin;
import com.android.build.gradle.LibraryExtension;

import org.gradle.api.Plugin;
import org.gradle.api.Project;

public class TimeLogPlugin implements Plugin<Project> {

    private static final String EXT_NAME = "timeLog";

    @Override
    public void apply(Project project) {
        boolean isApp = project.getPlugins().hasPlugin(AppPlugin.class);
        project.getExtensions().create(EXT_NAME, TimeLogConfig.class);
        project.afterEvaluate(project1 -> {
            TimeLogConfig config = (TimeLogConfig) project1.getExtensions().findByName(EXT_NAME);
            if (config == null) {
                config = new TimeLogConfig();
            }
            config.transform();
        });
        if (isApp) {
            AppExtension appExtension = project.getExtensions().getByType(AppExtension.class);
            appExtension.registerTransform(new TimeLogAppTransform());
            return;
        }
        if (project.getPlugins().hasPlugin("com.android.library")) {
            LibraryExtension libraryExtension = project.getExtensions().getByType(LibraryExtension.class);
            libraryExtension.registerTransform(new TimeLogLibraryTransform());
        }
    }
}

File: Plugin/TimeLogPligin/src/main/java/com/johnson/asm/timelog/TimeLogPlugin.kt
package com.johnson.asm.timelog

import com.android.build.gradle.AppExtension
import com.android.build.gradle.AppPlugin
import com.android.build.gradle.LibraryExtension
import com.johnson.asm.timelog.TimeLogConfig
import com.kronos.plugin.base.Log
import org.gradle.api.Plugin
import org.gradle.api.Project

class TimeLogPlugin : Plugin<Project> {
    override fun apply(project: Project) {
        val isApp = project.plugins.hasPlugin(AppPlugin::class.java)
        project.extensions.create(EXT_NAME, TimeLogConfig::class.java)
        project.afterEvaluate { project1: Project ->
            var config = project1.extensions.findByName(EXT_NAME) as TimeLogConfig?
            if (config == null) {
                config = TimeLogConfig()
            }
            //timeLog
            config.transform()
        }
        Log.info(">>> TimeLogConfig enable : ${TimeLogConfig.enable} TimeLogConfig filterTime : ${TimeLogConfig.filterTime}")
        if (!TimeLogConfig.enable) return
        if (isApp) {
            val appExtension = project.extensions.getByType(
                AppExtension::class.java
            )
            appExtension.registerTransform(TimeLogAppTransform())
            return
        }
        if (project.plugins.hasPlugin("com.android.library")) {
            val libraryExtension = project.extensions.getByType(
                LibraryExtension::class.java
            )
            libraryExtension.registerTransform(TimeLogLibraryTransform())
        }
    }

    companion object {
        private const val EXT_NAME = "timeLog"
    }
}